src/experimental/information_estimates.cpp

```diff
@@ -0,0 +1,407 @@
+                kt_double limit_y = cell.GetY() * m_cell_resol;
+
+                std::pair<std::vector<kt_double>, std::vector<kt_double>> intersections = utils::grid_operations::computeLineBoxIntersection(robot_pose.GetPosition(), laser_readings[i], robot_grid, beam_grid, limit_x, limit_y, m_cell_resol);
+
+                if (intersections.first.size() == 0)
+                    continue;
+
+                // Enter (d1) and Exit (d2) distances
+                std::vector<kt_double> distances;
+                for (int k = 0; k < intersections.first.size(); ++k)
+                {
+                    // From robot position to intersection points
+                    karto::Vector2<kt_double> intersection{intersections.first[k], intersections.second[k]};
+                    kt_double distance = robot_pose.GetPosition().Distance(intersection);
+                    distances.push_back(distance);
+                }
+
+                // Measurement outcomes vector {Pfree, Pocc, Pun}
+                std::vector<kt_double> probabilities {
+                    calculateScanMassProbabilityBetween(distances[1], m_max_sensor_range),
+                    calculateScanMassProbabilityBetween(distances[0], distances[1]),
+                    calculateScanMassProbabilityBetween(0.0f, distances[0])
+                };
+
+                // Appending new measurement outcomes for the current cell
+                appendCellProbabilities(probabilities, cell);
+
+                // Get all the measurement outcomes for the current cell
+                std::vector<std::vector<kt_double>> cell_prob = retrieveCellProbabilities(cell);
+
+                // Compute all the possible combinations for the current cell - algorithm 1
+                std::unordered_map<map_tuple, kt_double, utils::tuple_hash::HashTuple> meas_out_prob = computeMeasurementOutcomesHistogram(cell_prob);
+                
+                kt_double cell_mutual_inf = 0.0f;
+                for (auto& pair : meas_out_prob)
+                {
+                    cell_mutual_inf +=  pair.second * measurementOutcomeEntropy(pair.first);
+                }
+
+                // Mutual information of cell x, y given a set of measurements                
+                updateCellMutualInformation(1.0 - cell_mutual_inf, cell);
+            }
+        }
+        kt_double map_mut_info = calculateMapMutualInformation(); 
+        
+        // Extract the mutual information provided by this laser scan
+        kt_double laser_mut_info = calculateLaserMutualInformation(map_mut_info, m_curr_mut_info);
+        m_curr_mut_info = map_mut_info;
+
+        // Compare to assign the minimal value in the index and the mutual information.
+        if ((laser_mut_info < mut_info) && (laser_mut_info != 0.0))
+        {
+            // This is the assignation of the result
+            mut_info = laser_mut_info;
+            min_idx = curr_idx;
+        }
+        ++curr_idx;
```

Comment:
@kmilo7204 I don't understand this bit. IIUC information gain for a measurement `Z[0]` is `I(M, Z) - I(M, Z \ {Z[0]})`. You seem to be computing `I(M, {Z[0],...,Z[n]}) - I(M, {Z[0],...,Z[n - 1]}` instead.